src/core/config.js

```diff
@@ -0,0 +1,43 @@
+'use strict'
+
+const Joi = require('joi').extend(require('joi-multiaddr'))
+
+const schema = Joi.object().keys({
+  repo: Joi.alternatives().try(
+    Joi.object(), // TODO: schema for IPFS repo
+    Joi.string()
+  ).allow(null),
+  init: Joi.alternatives().try(
+    Joi.boolean(),
+    Joi.object().keys({ bits: Joi.number().integer() })
+  ).allow(null),
+  start: Joi.boolean(),
+  pass: Joi.string().allow(''),
+  EXPERIMENTAL: Joi.object().keys({
+    pubsub: Joi.boolean(),
+    sharding: Joi.boolean(),
+    dht: Joi.boolean()
+  }).allow(null),
+  config: Joi.object().keys({
+    Addresses: Joi.object().keys({
+      Swarm: Joi.array().items(Joi.multiaddr().options({ convert: false })),
+      API: Joi.multiaddr().options({ convert: false }),
+      Gateway: Joi.multiaddr().options({ convert: false })
+    }).allow(null),
+    Discovery: Joi.object().keys({
+      MDNS: Joi.object().keys({
+        Enabled: Joi.boolean(),
+        Interval: Joi.number().integer()
+      }).allow(null),
+      webRTCStar: Joi.object().keys({
+        Enabled: Joi.boolean()
+      }).allow(null)
+    }).allow(null),
+    Bootstrap: Joi.array().items(Joi.multiaddr().IPFS().options({ convert: false }))
+  }).allow(null),
+  libp2p: Joi.object().keys({
+    modules: Joi.object().allow(null) // TODO: schemas for libp2p modules?
+  }).allow(null)
+}).options({ allowUnknown: true })
```

Comment:
Hey @alanshaw. Love this. Can you explain why we allow unknown options? Are we trying to be helpful for 
developers or open to external config options?

Reply:
I don't really have a strong opinion either way on this one, I guess it's just to be flexible...

Reply:
oh look, I forgot I already wrote a reason in the description!

> The idea is that the validation should be flexible and shouldn't block new features landing in JS-IPFS. Validation of values that are objects **allows for unknown keys** so that features can land, and validation for the config can be backfilled at a later date if needs be.

Reply:
👍